File: ben/prolog/P31-P41-arithmetic/p35.pl
% Determine the prime factors of a given positive integer.
% Construct a flat list containing the prime factors in ascending order.

prime_factors(1,[]).
prime_factors(N, [X|L]) :-
    next_prime(N, 2, X), 
    N1 is N / X,
    prime_factors(N1, L), !.

next_prime(N, X, X) :-
    R is N mod X, 
    R == 0.
next_prime(N, D, X) :-
    R is N mod D,
    R =\= 0, 
    D1 is D + 1, 
    next_prime(N, D1, X).

File: ben/haskell/Q31-41-arithmetic/p35.hs
-- Determine the prime factors of a given positive integer. Construct a flat list containing the prime factors in ascending order.

primeFactors :: Int -> [Int]
primeFactors n0
    | n0 < 2 = []
    | otherwise = p : (primeFactors n1)
        where p = nextPrime 2 n0
              n1 = div n0 p

nextPrime :: Int -> Int -> Int
nextPrime min n 
    | mod n min == 0 = min
    | otherwise = nextPrime (min + 1) n
